library/core/helpers/class.ssoutils.php

```diff
@@ -145,21 +149,62 @@ protected function consumeStateToken($stateToken) {
      *
      * @param array $stateTokenData
      * @param string $stateToken
+     * @param string $source For logging purposes, optionally pass which token is being validated.
      * @return bool true if the data is valid and false otherwise.
      */
-    protected function isStateTokenValid($stateTokenData, $stateToken) {
+    protected function isStateTokenValid($stateTokenData, $stateToken, $source = '') {
         // Validate expected data.
-        if (!is_array($stateTokenData) || empty($stateTokenData['stateToken']) || empty($stateTokenData['exp'])) {
+        if (!is_array($stateTokenData)) {
+            $this->logger->event(
+                'state_token',
+                $this->logger::ERROR,
+                'Missing stateTokenData',
+                ['source' => $source, 'stateTokenData' => $stateTokenData ?? []]
```

Comment:
If we're only in here because `$stateTokenData` isn't an array, does it make sense to coalesce the value to an array?